pkg/egresspolicy/egresspolicy.go

```diff
@@ -0,0 +1,149 @@
+	labelsToMatch := labels.Set(podInfo.labels)
+	nsLabels := labels.Set(podInfo.nsLabels)
+	for k, v := range nsLabels {
+		labelsToMatch[policy.JoinPath(k8sConst.PodNamespaceMetaLabels, k)] = v
+	}
+	for _, selector := range config.endpointSelectors {
+		if selector.Matches(labelsToMatch) {
+			return true
+		}
+	}
+	return false
+}
+
+// Parse takes in a CiliumEgressNATPolicy CR type from the api server and converts it
+// to internal representation of the egress nat policy. It also performs some sanity
+// checks in parsing the ip, cidr and the selectors.
+func Parse(cenp *v2alpha1.CiliumEgressNATPolicy) (*EgressPolicyConfig, error) {
+	var endpointSelectorList []api.EndpointSelector
+	var dstCidrList []*net.IPNet
+
+	allowAllNamespacesRequirement := slim_metav1.LabelSelectorRequirement{
+		Key:      k8sConst.PodNamespaceLabel,
+		Operator: slim_metav1.LabelSelectorOpExists,
+	}
+	name := cenp.ObjectMeta.Name
+
+	if name == "" {
+		return nil, fmt.Errorf("CiliumEgressNATPolicy must have a name")
+	}
+
+	for _, cidrString := range cenp.Spec.DestinationCIDRs {
+		_, cidr, err := net.ParseCIDR(string(cidrString))
+		if err != nil {
+			log.WithError(err).Warnf("CiliumEgressNATPolicy %v error parsing cidr", name)
+			return nil, err
+		}
+		dstCidrList = append(dstCidrList, cidr)
+	}
+
+	for _, egressRule := range cenp.Spec.Egress {
+		if egressRule.NamespaceSelector != nil {
+			prefixedNsSelector := egressRule.NamespaceSelector
+			matchLabels := map[string]string{}
+			// We use our own special label prefix for namespace metadata,
+			// thus we need to prefix that prefix to all NamespaceSelector.MatchLabels
+			for k, v := range egressRule.NamespaceSelector.MatchLabels {
+				matchLabels[policy.JoinPath(k8sConst.PodNamespaceMetaLabels, k)] = v
+			}
+			egressRule.NamespaceSelector.MatchLabels = matchLabels
+
+			// We use our own special label prefix for namespace metadata,
+			// thus we need to prefix that prefix to all NamespaceSelector.MatchLabels
+			for i, lsr := range egressRule.NamespaceSelector.MatchExpressions {
+				lsr.Key = policy.JoinPath(k8sConst.PodNamespaceMetaLabels, lsr.Key)
+				egressRule.NamespaceSelector.MatchExpressions[i] = lsr
+			}
+
+			// Empty namespace selector selects all namespaces (i.e., a namespace
+			// label exists).
+			if len(egressRule.NamespaceSelector.MatchLabels) == 0 && len(egressRule.NamespaceSelector.MatchExpressions) == 0 {
+				egressRule.NamespaceSelector.MatchExpressions = []slim_metav1.LabelSelectorRequirement{allowAllNamespacesRequirement}
+			}
+
+			endpointSelectorList = append(
+				endpointSelectorList,
+				api.NewESFromK8sLabelSelector("", prefixedNsSelector, egressRule.PodSelector))
+		} else if egressRule.PodSelector != nil {
+			endpointSelectorList = append(
+				endpointSelectorList,
+				api.NewESFromK8sLabelSelector("", egressRule.PodSelector))
+		} else {
+			return nil, fmt.Errorf("CiliumEgressNATPolicy cannot have both nil namespace selector and nil pod selector")
```

Comment:
Our design is that the ns selector and pod selector cannot both be nil. Any proposed change here?

Reply:
I am not following, the `fmt.Errorf("CiliumEgressNATPolicy cannot have both nil namespace selector and nil pod selector")` can still exist. What I am suggesting is that `EgressRule` should perform an **and** operation with the `NamespaceSelector` and `PodSelector`:
```golang
type EgressRule struct {
	// Selects Namespaces using cluster-scoped labels. This field follows standard label
	// selector semantics; if present but empty, it selects all namespaces.
	NamespaceSelector *slimv1.LabelSelector `json:"namespaceSelector,omitempty"`

	// This is a label selector which selects Pods. This field follows standard label
	// selector semantics; if present but empty, it selects all pods.
	PodSelector *slimv1.LabelSelector `json:"podSelector,omitempty"`
}
```

Reply:
That scenario is already done in the parsing earlier, where i combined the ns selector and pod selector
```
                             endpointSelectorList = append(
				endpointSelectorList,
				api.NewESFromK8sLabelSelector("", prefixedNsSelector, egressRule.PodSelector))
```
Are you suggesting it could be done in the CRD? I don't know how to do that.

Reply:
Oh I totally missed that line! Sorry about it!